### ISSUE TITLE
lint: file-scoped check that requires type definition before its method definitions

### ISSUE THREAD

**Issue by quasilyte** [go-critic/go-critic#488]
Opinionated check. :nerd_face:
File-scoped.

Detects method definitions that go before receiver type definition.

**Before:**
```go
func (r recv) Method() {}
type recv struct{}
```

**After:**
```go
type recv struct{}
func (r recv) Method() {}
```

**Comment by Quasilyte** [comment]
We're lacking FileVisitor and/or DeclVisitor that would be a good fit for this.
I'll fix that soon.

**Comment by Quasilyte** [comment]
The implementation can look like this:
Walk file declarations, looking and every `*ast.FuncDecl` and `*ast.GenDecl` with `token.TYPE` kind.
If we have a function with receiver of type `T` that was not previously been defined with `*ast.GenDecl`, issue a warning and mark `T` as one that already caused warning (to avoid duplicated warnings). One may use type name as a map key or use `[]types.Type` list and use `types.Identical` function to check for equality.
